File: src/actions/actions.ts
import {
  addExpenseDB,
  deleteExpenseDB,
  getListDB,
  getSummary,
  updateExpenseDB,
} from "../db/db.ts";
import type { Expense } from "../db/types.ts";
import {
  addExpenseCSV,
  deleteExpenseCSV,
  listCSV,
  summaryCSV,
} from "../csv/csv.ts";

export async function addExpense(
  description: string,
  amount: number,
  date?: string,
) {
  if (!description || !amount || amount <= 0) {
    console.log("description and amount must exist to continue");
    return;
  }

  if (!date) {
    date = new Date().toISOString().split("T")[0];
  } else {
    // YYYY-MM-DD or YYYY-MM-DDTHH:MM
    if (!date.match(/^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2})?$/)) {
      console.log("Date must be in the format YYYY-MM-DD or YYYY-MM-DDTHH:MM");
      return;
    }

    try {
      date = new Date(date).toISOString().split("T")[0];
    } catch {
      console.error("Invalid date");
      return;
    }
  }

  const isExpenseAdded = await addExpenseDB({ date, description, amount });
  if (isExpenseAdded) {
    console.log(`Expense added successfully`);
    return;
  }

  await addExpenseCSV({ date, description, amount });
}

/* log a list of expenses in the following format:
 * ID    Date          Description      Amount
 * 1     2024-12-19    description      $100
 * 2     2024-12-19    description 2    $200
 * */
export async function list() {
  const expenses = await getListDB();

  if (expenses) {
    const columnWidths: number[] = [];
    const header = {
      id: "ID",
      date: "Date",
      description: "Description",
      amount: "Amount",
    } as unknown as Expense;

    expenses.unshift(header);

    // set column widths
    expenses.forEach((row) => {
      columnWidths[0] = Math.max(columnWidths[0] ?? 0, String(row.id).length);
      columnWidths[1] = Math.max(columnWidths[1] ?? 0, String(row.date).length);
      columnWidths[2] = Math.max(
        columnWidths[2] ?? 0,
        String(row.description).length,
      );
      columnWidths[3] = Math.max(
        columnWidths[3] ?? 0,
        String(row.amount).length,
      );
    });

    expenses.forEach((expense, index) => {
      const isHeader = index === 0;
      const basicPadding = 3;

      /*
      ID    Date          Description              Amount
      1     2024-12-02    Argentina car tickets    $837
      */
      console.log(
        String(expense.id).padEnd(columnWidths[0] + basicPadding),
        expense.date.padEnd(columnWidths[1] + basicPadding),
        expense.description.padEnd(columnWidths[2] + basicPadding),
        (!isHeader ? "$" : "") +
          String(expense.amount).padEnd(columnWidths[3] + basicPadding),
      );
    });

    return;
  }

  await listCSV();
}

/* log the total expenses in the following format:
 * `Total expenses: $30`
 * `Total expenses for August: $20`
 * */
export async function summary(month?: number) {
  if (month && (month < 1 || month > 12)) {
    console.log("Month must be between 1 and 12");
    return;
  }

  const totalAmount = await getSummary(month);
  console.log("total", totalAmount);

  if (totalAmount) {
    let monthName = "";
    if (month) {
      const currentYear = new Date().getFullYear();
      monthName = new Intl.DateTimeFormat("en", {
        month: "long",
      }).format(new Date(currentYear, month - 1));
    }

    console.log(
      `Total expenses${!month ? "" : ` for ${monthName}`}: $${totalAmount}`,
    );
    return;
  }

  await summaryCSV(month);
}

export async function deleteExpense(id: number) {
  if (!id || id < 1) {
    console.log("ID must be greater than 0");
    return;
  }

  const isExpenseDeleted = await deleteExpenseDB(id);
  if (isExpenseDeleted) {
    console.log("Expense deleted successfully");
    return;
  }

  await deleteExpenseCSV(id);
}

export async function updateExpense(expense: Expense) {
  if (!expense.id || expense.id < 1) {
    console.log("ID must be greater than 0");
    return;
  }

  const isExpenseUpdated = await updateExpenseDB(expense);
  if (isExpenseUpdated) {
    console.log("Expense updated successfully");
    return;
  }

  console.log("Expense not found and not updated");
}

File: src/actions/actions.py
import csv
from calendar import month_name
from datetime import datetime
from os import getenv
from typing import Optional

from dotenv import load_dotenv

dotenv_path = load_dotenv()
DB_PATH = getenv("DB_EXPENSE_TRACKER", "db.csv")


def add_expense(description: str, amount: float, date: Optional[str]) -> None:
    if not description or not amount or amount <= 0:
        raise Exception("description and amount must exist to continue")

    if date is None:
        date = str(datetime.today().date())

    with open(DB_PATH) as file:
        reader = csv.reader(file)
        rows = list(reader)

    last_expense = rows[-1]

    try:
        last_expense_id = int(last_expense[0])
    except ValueError:
        last_expense_id = 0

    new_id = last_expense_id + 1
    new_expense = [new_id, date, description, amount]

    with open(DB_PATH, "a", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(new_expense)
        print(f"Expense added successfully (ID: {new_id})")


# /* log a list of sorted expenses by date in the following format:
#  * ID    Date          Description      Amount
#  * 2     2024-12-19    description      $100
#  * 1     2024-12-20    description 2    $200
#  * */
def show_expenses() -> None:
    with open(DB_PATH) as file:
        reader = csv.reader(file)
        rows = list(reader)

    # set column widths
    column_width: list[int] = []
    for row in rows:
        for index, cell in enumerate(row):
            try:
                current_column_width = column_width[index]
            except IndexError:
                current_column_width = 0

            max_column_width = max(current_column_width, len(cell))
            if len(column_width) - 1 < index:
                column_width.extend([max_column_width])
            else:
                column_width[index] = max_column_width
    # sort by date
    data_rows = sorted(rows[1:], key=lambda row: datetime.fromisoformat(row[1]))
    sorted_rows = [rows[0]] + data_rows

    formatted_rows = []
    for row_index, row in enumerate(sorted_rows):
        for cell_index, cell in enumerate(row):
            is_amount_cell = cell_index == len(row) - 1
            basic_padding = 0 if is_amount_cell else 3
            amount_sign = "$" if is_amount_cell and row_index != 0 else ""
            padding = " " * (column_width[cell_index] + basic_padding - len(cell))

            text = f"{amount_sign}{cell}{padding}"
            if len(formatted_rows) - 1 < row_index:
                formatted_rows.extend([text])
            else:
                formatted_rows[row_index] += text

    for f_row in formatted_rows:
        print(f_row)


# /* log the total expenses in the following format:
#  * `Total expenses: $30`
#  * `Total expenses for August: $20`
#  * */
def summary(month: Optional[int]) -> None:
    total = 0.0

    with open(DB_PATH) as file:
        d_reader = csv.DictReader(file)
        rows = list(d_reader)

    if month:
        month_str = f"{month:02d}"
        rows = [row for row in rows if row["Date"].split("-")[1] == month_str]

    total = sum(float(row["Amount"]) for row in rows)

    month_info = f" for {month_name[month]}" if month is not None else ""
    print(f"Total expenses{month_info}: ${total}")


def delete_expense(id: int) -> None:
    with open(DB_PATH) as file:
        d_reader = csv.DictReader(file)
        d_rows = list(d_reader)
        id_found = False

        for index, row in enumerate(d_rows):
            if int(row["ID"]) == id:
                id_found = True
                del d_rows[index]
                break

        if not id_found:
            raise Exception("Expense with id not found")

    with open(DB_PATH, "w", newline="") as file:
        d_writer = csv.DictWriter(file, ["ID", "Date", "Description", "Amount"])
        d_writer.writeheader()
        d_writer.writerows(d_rows)
        print("Expense deleted successfully")


def update_expense(
    id: int, description: Optional[str], amount: Optional[float], date: Optional[str]
) -> None:
    with open(DB_PATH) as file:
        d_reader = csv.DictReader(file)
        d_rows = list(d_reader)
        id_found = False

        for row in d_rows:
            if int(row["ID"]) == id:
                id_found = True

                row["Description"] = description if description else row["Description"]
                row["Amount"] = amount if amount else row["Amount"]
                row["Date"] = date if date else row["Date"]
                break

        if not id_found:
            raise Exception("Expense not found and not updated")

    with open(DB_PATH, "w", newline="") as file:
        d_writer = csv.DictWriter(file, ["ID", "Date", "Description", "Amount"])
        d_writer.writeheader()
        d_writer.writerows(d_rows)
        print("Expense updated successfully")
